nni/mutable/mutable.py

```diff
@@ -0,0 +1,482 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+from __future__ import annotations
+
+__all__ = [
+    'Mutable', 'LabeledMutable', 'MutableSymbol', 'MutableExpression', 'Sample',
+    # 'Discrete', 'DiscreteMultiple', 'Continuous',
+]
+
+import copy
+import itertools
+import logging
+from typing import TypeVar, Type, Any, Generic, Dict, Iterable, Callable, List, TYPE_CHECKING, cast
+
+import numpy as np
+from numpy.random import RandomState
+from scipy.stats import norm, lognorm, uniform, loguniform
+
+from .exception import SampleValidationError, SampleMissingError
+from .symbol import SymbolicExpression, Symbol
+from .utils import auto_label
+
+if TYPE_CHECKING:
+    from scipy.stats import _distn_infrastructure
+
+Sample = Dict[str, Any]
+Choice = TypeVar('Choice')
+
+MISSING = '__missing__'
+
+T = TypeVar('T')
+
+_logger = logging.getLogger(__name__)
+
+
+def _is_mutable_symbol(mutable: Mutable) -> bool:
+    """Check if a mutable is a mutable symbol.
+
+    The default implementation of ``is_leaf``.
+    """
+    return isinstance(mutable, LabeledMutable)
+
+
+def dedup_labeled_mutables(mutables: Iterable[LabeledMutable]) -> dict[str, LabeledMutable]:
+    """Deduplicate mutables based on labels, and reform a dict.
+
+    Mutables are considered equal if they have the same label.
+    We will also check whether mutables of the same label are equal, and raise an error if they are not.
+
+    Parameters
+    ----------
+    mutables
+        The mutables to deduplicate.
+
+    Returns
+    -------
+    A dict. Keys are labels, values are mutables.
+    """
+    rv = {}
+    for mutable in mutables:
+        if not hasattr(mutable, 'label'):
+            raise ValueError('Mutable %s does not have a label' % mutable)
+        if mutable.label not in rv:
+            rv[mutable.label] = mutable
+        else:
+            if not mutable.equals(rv[mutable.label]):
+                raise ValueError(
+                    f'Mutables have the same label {mutable.label} but are different: {mutable} and {rv[mutable.label]}'
+                )
+    return rv
+
+
+class Mutable:
```

Comment:
Suggest more explicitly document which methods are to be used by "users", which are to be overridden by subclasses, and which are used by both.

Reply:
I list the "must-override" list. There is no should-suggestion. Subclass can do whatever they want.